### PR TITLE
Fix ucx compilation with gcc10

### PR DESCRIPTION
**Pull request by shuki-zanyovka**
## What
Fix UCX compilation with gcc10.

## Why ?
Fix UCX compilation with gcc10.

## How ?
Fix UCX compilation with gcc10.

### CODE REVIEW

**Comment by shuki-zanyovka** [comment]
> Can one of the admins verify this patch?

Adding @alex--m .

File: src/ucs/datastruct/ptr_array.h [hunk 1]
```diff
@@ -203,11 +203,12 @@ __ucs_ptr_array_for_each_get_step_size(ucs_ptr_array_t *ptr_array,
...
 
-    if (ucs_unlikely(__ucs_ptr_array_is_free((ucs_ptr_array_elem_t)elem))) {
-       size_elem = (elem >> UCS_PTR_ARRAY_FREE_AHEAD_SHIFT);
-    } else {
+    size_elem = (elem >> UCS_PTR_ARRAY_FREE_AHEAD_SHIFT);
+    if (ucs_likely(!__ucs_ptr_array_is_free((ucs_ptr_array_elem_t)elem))) {
        size_elem = 1;
     }
+    /* Prefetch the next item */
+    ucs_prefetch(&ptr_array->start[element_index + size_elem]);
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_array.h:211]: unrelated

Comment (shuki-zanyovka) [reply src/ucs/datastruct/ptr_array.h:211]: Removed.

Comment (shuki-zanyovka) [reply src/ucs/datastruct/ptr_array.h:211]: Sorry not removed, rechecking.

Comment (shuki-zanyovka) [reply src/ucs/datastruct/ptr_array.h]: Fixed.

File: src/ucs/sys/compiler_def.h [hunk 2]
```diff
@@ -135,7 +137,7 @@
  * @return Address of the container structure.
  */
 #define ucs_container_of(_ptr, _type, _member) \
-    ( (_type*)( (char*)(void*)(_ptr) - ucs_offsetof(_type, _member) )  )
+    ( (_type*)( (char*)(void*)((uintptr_t)(_ptr) - ucs_offsetof(_type, _member)) )  )
```
Comment (yosefe) [inline src/ucs/sys/compiler_def.h:140]: can we remote `(char*)` part?

Comment (shuki-zanyovka) [reply src/ucs/sys/compiler_def.h:140]: I believe we can, trying to figure out the issue below. Any idea what causes it?

@azure-pipelines
UCX PR (Build Static checks) Failing after 17m — Build Static checks failed

Comment (shuki-zanyovka) [reply src/ucs/sys/compiler_def.h:140]: @yosefe , Seems like the issue below isn't related to this commit:
@azure-pipelines
UCX PR (Build Static checks) Failing after 17m — Build Static checks failed

The issue seems to be the fact that no NULL checking is performed on the qp and qp->verbs.rd pointers.

----- issue ---
usr/bin/bash --noprofile --norc /__w/_temp/b4059e5d-6761-4931-b004-8d95518c4abc.sh
static checkers found errors:
Error: CLANG_WARNING:
/src/uct/ib/mlx5/ib_mlx5.c:12: included_from: Included from here.
/src/uct/ib/mlx5/ib_mlx5.inl:532:26: warning: The left operand of '!=' is a garbage value
#    if (qp->verbs.rd->pd != NULL) {
#                         ^

Error: CLANG_WARNING:
/src/uct/ib/rc/accel/rc_mlx5_common.c:943:33: warning: The left operand of '!=' is a garbage value
#    ucs_assert(iface->dm.dm->dm != NULL);
#               ~~~~~~~~~~~~~~~~ ^

**Comment by shamisp** [comment]
It would be nice to have GCC10 in CI

File: src/ucs/sys/compiler_def.h (hunk 2 above)
Comment (yosefe) [reply src/ucs/sys/compiler_def.h]: @shuki-zanyovka did you see such error on any other PR?

Comment (shuki-zanyovka) [reply src/ucs/sys/compiler_def.h]: No, it's somehow related to this PR, not sure what's the reason.

**Comment by shuki-zanyovka** [comment]
Closing, the following PR fixes the issue,
UCS: fix GCCv10 issue in queue for-each #5334
